system-test/spanner.ts

```diff
@@ -677,7 +677,7 @@ describe('Spanner', () => {
     it('should list the instances', done => {
       spanner.getInstances((err, instances) => {
         assert.ifError(err);
-        assert(instances.length > 0);
+        assert(instances!.length > 0);
```

Comment:
I wonder whether we need to consider this PR as a breaking change, as it does require users to make changes to their code in some cases. The callback for `getInstances` was untyped. It is now typed, and the `instances` field is declared as optional as it is not filled when we return an error. This in turn requires users to change their code to verify that `instances` is actually defined.
@skuruppu and @hengfengli WDYT?

Reply:
I agree that it should be considered as a breaking change if it requires users to make changes on their existing code. 

> This in turn requires users to change their code to verify that `instances` is actually defined.

When the callback for `getInstances` was untyped, I think users also need to check if `instances` is defined, but this is not forced by the type checker. It is based on the convention about how the callback is defined, `callback(err)` for a failure and `callback(null, results)` for a success. 

Now, due to the typings, it forces users to do this check, otherwise the type checker will give an error. Am I right?

Reply:
> Now, due to the typings, it forces users to do this check, otherwise the type checker will give an error. Am I right?

Correct.

Reply:
I'm happy with this breaking change if it helps users to make fewer mistakes in the future. @bcoe, what is the process for releasing breaking changes in the client libraries? Is it unacceptable, or is it ok to do so as long as we make it very clear?

Reply:
Just an fyi, I think there a major bump was already planned when support for Node 8 is dropped. You may just want to piggyback off of that.